beacon-chain/sync/backfill/status.go

```diff
@@ -0,0 +1,120 @@
+package backfill
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	types "github.com/prysmaticlabs/eth2-types"
+	"github.com/prysmaticlabs/prysm/beacon-chain/core/helpers"
+	"github.com/prysmaticlabs/prysm/beacon-chain/db"
+	"github.com/prysmaticlabs/prysm/proto/prysm/v1alpha1/block"
+)
+
+func NewStatus(store BackfillDB) *Status {
+	return &Status{
+		store: store,
+	}
+}
+
+type Status struct {
+	start       types.Slot
+	end         types.Slot
+	store       BackfillDB
+	genesisSync bool
+}
+
+// SlotCovered uses StartGap() and EndGap() to determine if the given slot is covered by the current chain history.
+// If the slot is beyond the current slot, an error will be returned.
+// If the slot is <= StartGap(), or >= EndGap(), the result is true.
+// If the slot is between StartGap() and EndGap(), the result is false.
+func (s *Status) SlotCovered(sl types.Slot) bool {
+	// short circuit if the node was synced from genesis
+	if s.genesisSync {
+		return true
+	}
+	if s.StartGap() < sl && sl < s.EndGap() {
+		return false
+	}
+	return true
+}
+
+// StartGap returns the slot at the beginning of the range that needs to be backfilled.
+func (s *Status) StartGap() types.Slot {
+	return s.start
+}
+
+// EndGap returns the slot at the end of the range that needs to be backfilled.
+func (s *Status) EndGap() types.Slot {
+	return s.end
+}
+
+var ErrAdvancePastOrigin = errors.New("cannot advance backfill Status beyond the origin checkpoint slot")
+
+// Advance advances the backfill position to the given slot & root.
+// It updates the backfill block root entry in the database,
+// and also updates the Status value's copy of the backfill position slot.
+func (s *Status) Advance(ctx context.Context, upTo types.Slot, root [32]byte) error {
+	if upTo > s.end {
+		return errors.Wrapf(ErrAdvancePastOrigin, "advance slot=%d, origin slot=%d", upTo, s.end)
+	}
+	s.start = upTo
+	return s.store.SaveBackfillBlockRoot(ctx, root)
+}
+
+// Reload queries the database for backfill status, initializing the internal data and validating the database state.
+func (s *Status) Reload(ctx context.Context) error {
+	_, err := s.store.GenesisBlockRoot(ctx)
+	if err != nil {
+		if errors.Is(err, db.ErrNotFoundGenesisBlockRoot) {
+			return errors.Wrap(err, "genesis block root required for checkpoint sync")
+		}
+		return err
+	}
+
+	cpRoot, err := s.store.OriginCheckpointBlockRoot(ctx)
+	if err != nil {
+		// mark genesis sync and short circuit further lookups
+		if errors.Is(err, db.ErrNotFoundOriginBlockRoot) {
+			s.genesisSync = true
+			return nil
+		}
```

Comment:
I find this a bit confusing. Shouldn't `originBlock == genesisBlock` when we sync from origin?